Рефакторинг MethodExecutor

Anton3 opened the pull request:
1. `MethodExecutor` теперь бросает исключения
2. `MethodExecutor` теперь всегда добавляет к запросу `accessToken`
3. `MethodExecutor` больше не провоцирует создание множества уровней
вложенности
4. Основная работа `MethodExecutorImpl` перенесена в `executeApiCall()`
5. Бросается напрямую `VkError`

Со старым `MethodExecutor` было тяжело и вообще непонятно как работать. Он мог иногда бросать исключение, а иногда возвращать код ошибки. Непонятно было, прикрепит он токен или нет. На практике, при работе с `MethodExecutor`, мы всегда хотим, чтобы он бросал исключения (так как он в любом случае бросает исключения сериализации и исключения от `HttpClient`), и мы всегда хотим, чтобы он прикреплял токен (основной сервер ВК запросы без токена, вроде, не принимает).

Кроме того, такой интерфейс `MethodExecutor` готовит почву для добавления нормального `BatchMethodExecutor` с поддержкой ограничения частоты и с поддержкой метода execute (по 25 команд за раз).

P.S. Код, использующий (но не создающий) MethodExecutor, вроде, не сломается.

Anton3 committed:
Добавить плагин Kotlin Gradle

Anton3 committed:
Незначительные изменения в файлах сборки

Anton3 committed:
Новая версия MethodExecutor
1. MethodExecutor теперь бросает исключения
2. MethodExecutor теперь всегда добавляет к запросу accessToken
3. MethodExecutor больше не провоцирует создание множества уровней
вложенности
4. Основная работа MethodExecutorImpl перенесена в executeApiCall()
5. Бросается напрямую VkError

alatushkin commented:
1)Не вижу достаточных причин всегда бросать исключение когда VK присылает ошибку - т.к. ответ API не являет исключением, а является одним из ожидаемых результатов. 
При этом, не исключая кейсов когда такое поведение может быть удобным - есть возможность добавить такое поведение при необходимости глобально или "по месту использования" с помощью throwExceptionsOnError()

2)Зачем всегда добавлять accessToken? Есть кейсы в которых MethodExecuter конструируется "где-то", а токен биндится по месту использования. А благодаря тому что возможна функциональная композиция на него может быть навешано разное поведение (тротлинг, управление rps в зависимости от БЛ связанной с токенами и прочее что душе угодно, без необходимости менять сам executer), список которого нет необходимости раскрывать там где речь заходит о токене. Грубо говоря, с непривязанным токеном он работает как передача замыкания, фабрики.
При этом - снова есть по логике в приложении всегда один и тот же токен - его можно сразу и привязать через withToken

3)Боюсь, не совсем понял твою мысль про "провоцирование уровней сложности", что именно это означает и почему это плохо.
4)В чём выигрыш замены interface{ operator invoke} на fun?
5)Объект ошибки api != объект исключение.  Можно насильственно их смешать, но зачем?

P.S Выполнять по 25 команд за раз можно будет не меняя этот executer

Anton3 commented:
`1.` Хотим мы этого или нет, но MethodExecutor в любом случае бросает исключения:

- IOException, когда HttpClient не достучался до сервера
- Исключения парсинга JSON, когда ВК отдаёт некорректный JSON (а это бывает довольно часто)

Сейчас обработка исключений в моём коде на официальном API выглядит как-то так:
```kotlin
try {
    return execute()
} catch (e: ApiException) {
    if (e.errorCode == NOT_FOUND) return null
    else throw e
} catch (e: JSONException) {
    log.warn("Incorrect JSON!")
} catch (e: IOException) {
    log.warn("Network troubles")
}
```
Обработка ошибок по текущему варианту kotlin-vk-api выглядит так:
```kotlin
try {
    val result = execute()
    if (result.error != null) {
        if (result.error.errorCode == NOT_FOUND) return null
        else throw e
    }
    return result.response!!
} catch (e: JSONException) {
    log.warn("Incorrect JSON!")
} catch (e: IOException) {
    log.warn("Network troubles")
}
```
Это помесь исключений и кодов ошибок. Не надо так. Плюс, коды ошибок и "ручная" обработка ошибок могли быть актуальны в C/C++, но не являются идиоматичными в Kotlin (сравните: монада CompletableFuture заменяется в Kotlin на suspend функции).

Интерфейс `MethodExecutor` от этого упростится: вместо "возвращает результат, или бросает исключение, или возвращает код ошибки" будет "возвращает результат, или бросает исключение".

`2.` Проблема в том, что все API методы (остальные `MethodExecutor` не обрабатывает) требуют MethodExecutor с привязанным токеном. Что будет, если передавать в какую-нибудь функцию MethodExecutor без токена? ~~Краш~~ Бесконечные ошибки. У нас статически типизированный язык, решить эту проблему проще простого - разделить интерфейсы на `MethodExecutorWithToken` и `MethodExecutorWithoutToken`, и чтобы использовать первый, надо вначале создать второй.

То же самое другими словами: для пользователя API информация о том, прикрепляется ли токен, является критически важной, поэтому её нельзя абстрагировать. 

Чтобы не плодить интерфейсы, пусть `MethodExecutorWithToken` будет простым классом-прокси к `MethodExecutorWithoutToken`.

Сомневаюсь насчёт имени этих сущностей:

Без токена | С токеном
---:|:---
MethodExecutorWithoutToken | MethodExecutorWithToken
RawMethodExecutor | MethodExecutor
MethodExecutor | VkClient
… | …

`3.` Не важно, я сам себя не очень понял :D
`4.` Выигрыша никакого. Это небольшой рефакторинг кода `MethodExecutorImpl`
`5.` Если делать отдельное исключение, то интерфейс этого исключения надо дополнительно развивать, создавать отдельные классы исключений - как в официальном API. Нужно нам это? Мне кажется, проще сделать `VkError : Exception`, чтобы избежать бесконечного заворачивания-доставания кода ошибки.

Anton3 committed:
Разделить MethodExecutor на MethodExecutor и VkClient
1. MethodExecutor не владеет токеном, VkClient владеет
2. Удалил executeApiCall, теперь это снова деталь реализации

Anton3 committed:
Разнести `MethodExecutor` и `VkClient` по разным файлам

alatushkin commented:
1)Подход "результат или код ошибки" - весьма распространен и вне C/C++. Еще Python,Js,Go и наверняка куча других языков. И даже в Котлине - через общий sealed class у которого потомки - и классы с результатом и класс с ошибкой (и я думал о таком варианте, но тут я уже подустал с генератором сорсов и решил не делать). 
+ответ api - не исключение.

Но в сухом остатке - речь только  о том что для логики твоего кода это неважно и и то и другое "исключения" и тебе неудобно это описывать в одном месте 2 способами  - аргумент принимается. Но и тут ничего переделывать не нужно:
Когда сконструировал MethodExecuter (с httpClient, токеном и прочими декораторами) - в конце вызови throwExceptionsOnError - и получишь fun <T> invoke(method: VkMethod<T>): T - он будет сразу возвращать результат или выбрасывать исключение, если случилась ошибка VK.  
И в коде используй уже его. О том в этом вопросе бывают разные подходы я уже подумал и предусмотрел.

2)Аргумент верен, но лишь от части. Да - если не вызвать .withToken при конструировании перед пбуликацией объекта либо перед непосредственным вызовом - действительно vk вернет ошибку мол токена нет. Но я не вижу чтобы это было проблемой:
+ Можно в сам executer добавить assert чтобы он даже не пытался выполнять запросы без токена и выдавал соответсвующее сообщение.
+ можно сделать одну/несколько фабричных функций (HttpClient)->Executer  , (HttpClient,token)->Executer которые бы помогали не забыть про биндинг токена, для тех кому сложно
+ Ошибка при неправильном использовании библиотеки, которую легко исправить добавив вызов функции в одном месте - не то что бы серезная проблема, что бы что-то менять. Так с любой библиотекой - если ты ее используешь неправильно - случается ошибка.
+ Если продолжать логику что всё же это среьзная проблема требующая изменений - то если подумать еще немного - выяснится, что проблема неправильной работы с токенами всё-равно не решена - мы не можем полагаться на компилятор в вопросе проверки типа токена - ведь разные методы могут быть вызваны только с токенами определенного типа.

5)В тех кейсах о которых ты говоришь - между if(e.vkError.code) и if(e.code) - разница не велика.
При этом VkError остается "неисключением" и продолжает использоваться в логике (result,error)
Не теряю надежды чуть позже дописать генератор и выдать sealed class иерархию под VkError чтобы можно было смотреть не на коды ошибок а на мнемонику имени класса ошибки
И соответсвующую ей иерархию VkErrorException

alatushkin commented:
По итогу, я вижу, что тебе не всё-равно и хочется внести вклад в улучшение и это оч круто, но в данном случае ты направляешь свои усилия совсем не на то, что принесет бОльшую пользу, а на в большей мере вкусовые правки. При том что даже в существующем виде библиотека вполне позволяет тебе писать твой код так как ты привык (просто не забудь про withToken и throwExceptionsOnError при конструировании и ты получишь то, поведение которое тебе нужно)

Anton3 commented:
`1.` Внезапно понял, что `MethodExecutorWithException` не наследуется от `MethodExecutor`. Посыпаю голову пеплом. Проблема в том, что `MethodExecutorWithException` должен быть не `interface`, а `class`, так как расширять именно его функциональность не имеет смысла. Это меня и смутило. Плюс, `MethodExecutorWithException` должен быть классом "по умолчанию" в клиентском коде, поэтому его имеет смысл назвать `VkClient`.

`2.` То, что `MethodExecutor` с токеном и без токена одного класса, нарушает [LSP](https://ru.wikipedia.org/wiki/Принцип_подстановки_Барбары_Лисков). А именно: 99% кода, принимающего `MethodExecutor`, на самом деле хочет не `MethodExecutor`, а `MethodExecutorWithToken`. (Сейчас это ещё и недоступный анонимный класс, откуда не вытащить токен 🤷‍♂️.)

> Если продолжать логику что всё же это среьзная проблема требующая изменений - то если подумать еще немного - выяснится, что проблема неправильной работы с токенами всё-равно не решена - мы не можем полагаться на компилятор в вопросе проверки типа токена - ведь разные методы могут быть вызваны только с токенами определенного типа.

Вот и я думал, как эту статическую проверку типа токена сюда добавить. Это важная проблема, которая тоже требует решения. Подумаю ещё, что можно с этим сделать.

`5.` Согласен, что так будет лучше. А пока можно хотя бы добавить расширение в `VkErrorException` для получения кода ошибки. Хотя не понял, почему нужна отдельная иерархия для `VkError`, ведь этот класс должен содержать ровно то, что отдал сервер ВК.

И не понимаю, почему нужно наследовать `VkErrorException` от `RuntimeError`, а не от `IOException`, например. Тогда достаточно будет объявлять `@Throws(IOException::class)`, чтобы это работало с Java.

В общем, подготовлю новый чистенький PR, и возьмусь за Ktor.

alatushkin commented:
Про MethodExecutor с токеном и без - в одном типе или в разных - логика в том что ты говоришь есть: 
1) действительно - большая часть кейсов - работа идет с уже привязанным токеном. Без токена - оно работать не может.
2) действительно - в чем то даже принцип SRP нарушается: Executor работает и как исполнитель запросов и как фабрика по порождению новых версий себя (с привязанным токеном и или новым поведением).

Но это всё планировалось как некий набор внутренних реализаций в котором бы удобно было комбинировать поведение. Снаружи всё скрыто и выставлен только 1 интерфейс (method)->result а все инициализации - в helper-функции. Если туда передать токен - она сконструирует уже с привязанным токеном. 
А раз у MethodExecutor и VkClient - интерфейс один и тот же ( (method)->result ) - так ли нужен этот разный нейминг и раскрытие деталей реализации или всё же это может быть скрыто внутри хелпера/хелперов который из кубиков собирает нужное поведение и отдает наружу не раскрывая деталей? 

5)Про VkErrorException от RuntimeError, а не от IOException, например.
Тут логика такая: если этот код когда-то всё же выйдет в java мир - многие (в том числе и я) - не любят когда их вынуждают писать catch, только потому что у метода объявлен checked exception, и внутри catch - как правило просто оборачивает в другое исключение и бросается дальше. Потому и использовал RuntimeException - потому что оно unchecked


В общем про нейминг и устройство (с токеном/без, с выбрасыванием исключений/без) еще можно подумать, но лучше в виде снипетов кода - нет смысла перелопачивать всё и делать PR чтобы обсудить сигнатуру 2х методов в интерфейсе и пару строк кода примеров вызова

Anton3 commented:
> А раз у MethodExecutor и VkClient - интерфейс один и тот же ( (method)->result ) - так ли нужен этот разный нейминг и раскрытие деталей реализации или всё же это может быть скрыто внутри хелпера/хелперов который из кубиков собирает нужное поведение и отдает наружу не раскрывая деталей?

У них разный интерфейс с логической точки зрения, именно это нарушает [LSP](https://ru.wikipedia.org/wiki/Принцип_подстановки_Барбары_Лисков). Хотя интерфейс (method)->result одинаковый, но их нельзя заменять один на другой в одних и тех же местах (функция упадёт с исключением).  Поэтому я и говорю, что это должны быть 2 разных интерфейса: с токеном и без токена.

> Тут логика такая: если этот код когда-то всё же выйдет в java мир - многие (в том числе и я) - не любят когда их вынуждают писать catch, только потому что у метода объявлен checked exception, и внутри catch - как правило просто оборачивает в другое исключение и бросается дальше. Потому и использовал RuntimeException - потому что оно unchecked

Как мы уже выяснили, бросать и обрабатывать `IOException` придётся в любом случае. То есть текущий вариант уже некорректен, и должен быть `@Throws(IOException::class)`. Так почему бы не наследовать `VkErrorException` от `IOException`, чтобы и `VkErrorException` тоже обрабатывался корректно?

> В общем про нейминг и устройство (с токеном/без, с выбрасыванием исключений/без) еще можно подумать, но лучше в виде снипетов кода - нет смысла перелопачивать всё и делать PR чтобы обсудить сигнатуру 2х методов в интерфейсе и пару строк кода примеров вызова

Ок, если будет тяжко с Git, скину сниппеты))

Anton3 commented:
Ещё открытый вопрос: действительно ли так нужен "бросающий `MethodExecutor` без токена"?

alatushkin commented:
В этом и прелесть простых функциональных интерфейсов - он допускает разную логику при этом не ограничивая тех, кому она не нужна или нужна "другая")


P.S Там скоро будет обновление с возможностью за один вызов выполнять до 25 одинаковых методов: немного закопался т.к. у метода execute совершенно другой формат ответа. ума не приложу, как лучше "вписать"

Anton3 commented:
Есть наработки по типобезопасным токенам. Код ужасен, поэтому пока показывать не буду, а покажу примеры использования, чтобы понять, нужно ли нам вообще это.

Во-первых, нужно дополнить кодогенератор (я в нём не разбираюсь), чтобы все наследники `VkMethod<T>`, помимо него, наследовались ещё и от одного из следующих интерфейсов (в зависимости от допустимых типов токенов:

- `UserMethod`
- `ServiceMethod`
- `UserGroupMethod`
- `UserServiceMethod`
- `UserGroupServiceMethod`

Далее, вместо строки, в `withToken` надо будет передавать один из следующих классов:

- `data class UserToken(val accessToken: String, val id: Int)`
- `data class GroupToken(val accessToken: String, val id: Int)`
- `data class ServiceToken(val accessToken: String, val id: Int, val clientSecret: String)`

Пример:

![image](https://user-images.githubusercontent.com/2719663/49434752-fec8f180-f7c5-11e8-910b-93429a6e6c86.png)

Anton3 commented:
Отдельная просьба закоммитить локальные изменения, чтобы я сделал у себя pull и мог уже начать решать мерж конфликты.

alatushkin commented:
Все изменения уже в мастере. Локальных - нет
